### ISSUE TITLE
SpendingCondition: use ripemd160(sha3(bytecode)) instead of just using ripemd160

### ISSUE THREAD

**Issue by pinkiebell** [leapdao/leap-node#242]
# Bounty

## Scope

-
-
-

## Roles
bounty gardener: name / share
bounty worker: name / share
bounty reviewer: name / share

**Comment by troggy** [comment]
related? https://github.com/leapdao/leap-core/commit/832f2551ff5472c740af1b79939a51bff3979468#diff-e652b5ccb2708f628a9de325cf105fb7R262

**Comment by pinkiebell** [comment]
>  related? leapdao/leap-core@832f255#diff-e652b5ccb2708f628a9de325cf105fb7R262

Yes, related. But we still need to do the same for the addressing scheme.
@johannbarbie 
This one slipped through :/

**Comment by pinkiebell** [comment]
One proposal after all the madness today, and that I discovered that the solidity abi encoder doesn't hash the length of a dynamic `bytes` array:
better use `ripemd160(sha3(sha3(length), sha3(bytecode)))`

@johannbarbie What do you think? :)

**Comment by johannbarbie** [comment]
why not`ripemd160(sha3(uint256 length, sha3(bytecode)))`?

**Comment by johannbarbie** [comment]
ideally we would use `ripemd160(uint256 length, bytecodeMerkleRoot)`, so that solEVM can use the value and can be fed with compact proofs.

**Comment by troggy** [comment]
@johannbarbie I guess we should do it the right way before the migration, otherwise it would be a breaking change later, right?

**Comment by pinkiebell** [comment]
> ideally we would use `ripemd160(uint256 length, bytecodeMerkleRoot)`, so that solEVM can use the value and can be fed with compact proofs.

`ripemd160(keccak256(uint256 length, bytecodeMerkleRoot))` 😄

**Comment by troggy** [comment]
bytecodeMerkleRoot = merkleized array of opcodes? Do you have a code for it in solEVM already which could be reused?

**Comment by johannbarbie** [comment]
> Do you have a code for it in solEVM already which could be reused?

nope, but @peara working on it here: https://github.com/leapdao/solEVM-enforcer/issues/117

**Comment by troggy** [comment]
ah, I see it should be exactly the same as in solEVM

**Comment by troggy** [comment]
@johannbarbie @pinkiebell can someone point me out why we need length in here just to produce hash?

**Comment by pinkiebell** [comment]
@troggy This helps us securing the validation of proofs. If we know the length, we know the tree depth and therefore an attacker won't be able to send us bogus-length proofs for a 'theoretical' birthday attack

**Comment by pinkiebell** [comment]
@troggy 
Ah sorry, I was deep into solEVM territory, assume we want to validate the correctness for a given opcode(s) to be used in the verification game by using merkle proofs.
We gain more security by knowing the length of the bytecode of the contract and can therefore calculate the tree depth for the contract's bytecode merkle tree to enforce a strict size on the number proofs to be used to validate the submission.

**Comment by troggy** [comment]
From the Ops Tactical:

decided to go with `ripemd160(length, sha3(bytecode))` here and create another spending condition type once https://github.com/leapdao/solEVM-enforcer/issues/117 is done
